### ISSUE TITLE
OSD pod permissions broken, unable to open OSD superblock after node restart

### ISSUE THREAD

**Issue by t3hmrman** [rook/rook#7519]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**

`osd` pods fail to restart after node restart

**Expected behavior:**

OSD pods come back online

**How to reproduce it (minimal and precise):**

- Set up two pieces of storage -- a partition on one drive (in my example, one that holds the OS) and a full drive 
- Set up a cluster using `k0s` ([there are some caveats there](https://github.com/k0sproject/k0s/issues/810))
- Switch to ceph version 15.2.6 ([`15.2.9` does not work partitions are not properly considered by a `ceph volume`/`ceph batch` command being run](https://github.com/ceph/ceph/pull/38280))
- Install Rook 1.5.9
- Observe that the pieces all start correctly, PVCs create PVs, both are bound (though pods do not succeed because while attaching works, mounting fails, but that's a separate issue I'm trying to debug), block device is present in Ceph Dashboard
- Restart the node
- Observe that all pods in `rook-ceph` come up, except the OSD pods which are stuck (`CrashLoopBackoff`)
- Observe the following in the logs:

```console
$ k logs rook-ceph-osd-1-59765fcfcd-p5mp5
debug 2021-04-02T13:59:19.854+0000 7f76bce47f40  0 set uid:gid to 167:167 (ceph:ceph)
debug 2021-04-02T13:59:19.854+0000 7f76bce47f40  0 ceph version 15.2.6 (cb8c61a60551b72614257d632a574d420064c17a) octopus (stable), process ceph-osd, pid 1
debug 2021-04-02T13:59:19.854+0000 7f76bce47f40  0 pidfile_write: ignore empty --pid-file
debug 2021-04-02T13:59:19.854+0000 7f76bce47f40 -1 bluestore(/var/lib/ceph/osd/ceph-1/block) _read_bdev_label failed to open /var/lib/ceph/osd/ceph-1/block: (13) Permission denied
debug 2021-04-02T13:59:19.854+0000 7f76bce47f40 -1  ** ERROR: unable to open OSD superblock on /var/lib/ceph/osd/ceph-1: (2) No such file or directory
```

I thought this might be [related to the udev rule issue that came up in the runners](https://github.com/rook/rook/pull/7467/files) so I tried dropping in some udev rules (confirmed the the changed owner) and the errors persist. When I went to check the folder as the `ceph` user, it turns out the folder is empty:

```console
$ sudo su -l ceph -s /bin/bash
ceph@all-in-one-01:~$ tree /var/lib/ceph/
/var/lib/ceph/
├── bootstrap-mds
├── bootstrap-mgr
├── bootstrap-osd
├── bootstrap-rbd
├── bootstrap-rbd-mirror
├── bootstrap-rgw
├── crash
│   └── posted
├── mds
├── mgr
├── mon
├── osd
└── tmp

13 directories, 0 files
```

**Environment**:
* OS (e.g. from /etc/os-release):

```
NAME="Ubuntu"
VERSION="20.04.2 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.2 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
```

* Kernel (e.g. `uname -a`):

```
Linux all-in-one-01 5.4.0-67-generic #75-Ubuntu SMP Fri Feb 19 18:03:38 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

* Cloud provider or hardware configuration:

Hetzner, bare metal

* Rook version (use `rook version` inside of a Rook Pod):

```
1.5.9
```

* Storage backend version (e.g. for ceph do `ceph -v`):

```
15.2.6
```

* Kubernetes version (use `kubectl version`):

1.20

* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift):

bare metal (provisioned by `k0s`)

* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/master/ceph-toolbox.html)):

Dashboard is in HEALTH_WARN, but I assume they are benign for the following reasons:

- `PG_AVAILABILITY: Reduced data availability: 1 pg inactive`
  - This makes sense because all cluster itself is inactive , there are no PVs
- `PG_DEGRADED: Degraded data redundancy: 1 pg undersized`
  - This makes sense because 1 pg is actually "non HA" and does not have any redundancy
- `POOL_NO_REDUNDANCY: 1 pool(s) have no replicas configured`
  - See previous note about the "undersized" pg
- `TOO_FEW_OSDS: OSD count 2 < osd_pool_default_size 3`
  - There are only 2 things to put OSDs on (1 disk and 1 partition) in the whole "cluster", this makes sense.
  - I actually changed `osdsPerDevice` to 5 since these are NVMe disks

```
- PG_AVAILABILITY: Reduced data availability: 1 pg inactive
- PG_DEGRADED: Degraded data redundancy: 1 pg undersized
- POOL_NO_REDUNDANCY: 1 pool(s) have no replicas configured
- TOO_FEW_OSDS: OSD count 2 < osd_pool_default_size 3
```

With `osdsPerDevice` set to `"5"` all these warnings go away, except for the `POOL_NO_REDUNDANCY` (one of my pools is intentionally without redundancy), and OSDs start just fine *the first time* so I think these are indeed benign

**Comment by haslersn** [comment]
It seems there are now multiple problems that appear after node restarts: see #7497.

**Comment by t3hmrman** [comment]
Ah so I should have posted here but I've been working on this all day and just made what I *think* is a breakthrough in my own case, it looks like Ceph implicitly depends on some udev changes or GPT to change permissions for drives properly.

What's weird is that everything I see on the folders and disk states that the folder is owned by the `ceph` user (uid 167) that is trying to write to it (and if I `exec` into the containers I can write new files to the mount folder with both `root` and `ceph` users), but  [someone also had something similar happen way back in 2018 in `ceph/ceph-ansible`](https://github.com/ceph/ceph-ansible/issues/2354) and the [Ceph bugtracker has a ticket which was opened a year ago (2020) and hasn't been addressed yet](https://tracker.ceph.com/issues/41777).

I changed my setup to use GPT rather than trying to add the udev changes myself (which I actually did before going down a different rabbit hole), and that *might* have solved the problem, if the machine booted after Rook had finished setting up. I ended up with a partition layout like this:

```console
root@all-in-one-01 ~ # lsblk -f
NAME                                                                                                 FSTYPE      LABEL UUID                                   FSAVAIL FSUSE% MOUNTPOINT
nvme0n1                                                                                                                                                                       
├─nvme0n1p1                                                                                          swap              971ad85d-afb5-48ec-94d2-bfcc1db45d1e                  [SWAP]
├─nvme0n1p2                                                                                          ext4              02587490-1487-46c9-9443-f875a9bc7434    757.5M    15% /boot
├─nvme0n1p3                                                                                          ext4              747b604d-0a3e-4a23-8415-ab5295d1f6cb     51.1G    13% /
├─nvme0n1p4                                                                                          ext4              7f7ec944-1075-47ec-a4f2-8ee338757379                  
└─nvme0n1p5                                                                                                                                                                   
nvme1n1                                                                                              LVM2_member       syLoW1-pMwf-wHco-vNo9-hYGt-Pr4n-IECcmq                
├─ceph--987bbd35--383f--4c30--bf07--e48364cc84f7-osd--data--5e05a235--445c--4cad--a70a--fe5990fad4e2                                                                          
├─ceph--987bbd35--383f--4c30--bf07--e48364cc84f7-osd--data--d64ca26b--e0c2--4c66--83fd--df15e99eee31                                                                          
├─ceph--987bbd35--383f--4c30--bf07--e48364cc84f7-osd--data--5f86d5bc--f436--4c18--8921--74bcb1e1cb97                                                                          
├─ceph--987bbd35--383f--4c30--bf07--e48364cc84f7-osd--data--e9dfa596--5c85--49a9--9db5--023aee345949                                                                          
└─ceph--987bbd35--383f--4c30--bf07--e48364cc84f7-osd--data--cd436f2d--a95c--4bcb--a18e--8abff1577286 
```

So Ceph got the second disk (`nvme1n1`) and completely consumed it, adding 5 OSDs which got their own little space (LVM LVs). The first disk was ignored, but I think removal of the boot sector from that second disk might be the issue there. So if I'm stuck with MBR-based systems (since GPT isn't going to work out) then I have to craft udev rules.

There's lots of rumblings about this out there but I took [inspiration from an old gist with some Ceph flavored udev rules](https://gist.github.com/ChrisMacNaughton/887a5365960af097e809927d9b420625), but in the end I just went with trying to run `ceph-volume lvm activate --all` and that didn't work either. 

Absolutely stumped on this, I tried changing the `runAsUser` to `167` (the ceph user on my machine) and also giving more capabilities. 

When I go into the container and run it myself (swap out the deployment with `bash -c 'sleep infinity'`, etc) I see this:

```console
[root@rook-ceph-osd-9-78494cff76-xs5h5 /]# # ceph-osd --foreground --id "9" --fsid 6d0f3f65-0bd5-4e17-bee7-91c3cc494c5b --setuser ceph --setgroup ceph --crush-location="root=default host=all-in-one-01" --log-to-stderr=true --err-to-stderr=true --mon-cluster-log-to-stderr=true '--log-stderr-prefix=debug ' --default-log-to-file=false --default-mon-cluster-log-to-file=false --ms-learn-addr-from-peer=false
[root@rook-ceph-osd-9-78494cff76-xs5h5 /]# ceph-osd --foreground --id "9" --fsid 6d0f3f65-0bd5-4e17-bee7-91c3cc494c5b --setuser ceph --setgroup ceph --crush-location="root=default host=all-in-one-01" --log-to-stderr=true --err-to-stderr=tr
ue --mon-cluster-log-to-stderr=true '--log-stderr-prefix=debug ' --default-log-to-file=false --default-mon-cluster-log-to-file=false --ms-learn-addr-from-peer=false
debug 2021-04-04T08:20:25.425+0000 7fb55f4a6f40  0 set uid:gid to 167:167 (ceph:ceph)
debug 2021-04-04T08:20:25.425+0000 7fb55f4a6f40  0 ceph version 15.2.6 (cb8c61a60551b72614257d632a574d420064c17a) octopus (stable), process ceph-osd, pid 22
debug 2021-04-04T08:20:25.425+0000 7fb55f4a6f40  0 pidfile_write: ignore empty --pid-file
debug 2021-04-04T08:20:25.429+0000 7fb55f4a6f40  1 bdev create path /var/lib/ceph/osd/ceph-9/block type kernel
debug 2021-04-04T08:20:25.429+0000 7fb55f4a6f40  1 bdev(0x55bafd244380 /var/lib/ceph/osd/ceph-9/block) open path /var/lib/ceph/osd/ceph-9/block
debug 2021-04-04T08:20:25.429+0000 7fb55f4a6f40  1 bdev(0x55bafd244380 /var/lib/ceph/osd/ceph-9/block) open size 85022736384 (0x13cbc00000, 79 GiB) block_size 4096 (4 KiB) non-rotational discard supported
debug 2021-04-04T08:20:25.429+0000 7fb55f4a6f40  1 bluestore(/var/lib/ceph/osd/ceph-9) _set_cache_sizes cache_size 3221225472 meta 0.4 kv 0.4 data 0.2
debug 2021-04-04T08:20:25.429+0000 7fb55f4a6f40  1 bdev create path /var/lib/ceph/osd/ceph-9/block type kernel
debug 2021-04-04T08:20:25.429+0000 7fb55f4a6f40  1 bdev(0x55bafd244a80 /var/lib/ceph/osd/ceph-9/block) open path /var/lib/ceph/osd/ceph-9/block
debug 2021-04-04T08:20:25.429+0000 7fb55f4a6f40  1 bdev(0x55bafd244a80 /var/lib/ceph/osd/ceph-9/block) open size 85022736384 (0x13cbc00000, 79 GiB) block_size 4096 (4 KiB) non-rotational discard supported
debug 2021-04-04T08:20:25.429+0000 7fb55f4a6f40  1 bluefs add_block_device bdev 1 path /var/lib/ceph/osd/ceph-9/block size 79 GiB
debug 2021-04-04T08:20:25.429+0000 7fb55f4a6f40  1 bdev(0x55bafd244a80 /var/lib/ceph/osd/ceph-9/block) close
debug 2021-04-04T08:20:25.717+0000 7fb55f4a6f40  1 bdev(0x55bafd244380 /var/lib/ceph/osd/ceph-9/block) close
debug 2021-04-04T08:20:25.989+0000 7fb55f4a6f40  1  objectstore numa_node 0
debug 2021-04-04T08:20:25.989+0000 7fb55f4a6f40  0 starting osd.9 osd_data /var/lib/ceph/osd/ceph-9 /var/lib/ceph/osd/ceph-9/journal
debug 2021-04-04T08:20:25.989+0000 7fb55f4a6f40 -1 Falling back to public interface
debug 2021-04-04T08:20:25.993+0000 7fb55f4a6f40  0 load: jerasure load: lrc load: isa
debug 2021-04-04T08:20:25.993+0000 7fb55f4a6f40  1 bdev create path /var/lib/ceph/osd/ceph-9/block type kernel
debug 2021-04-04T08:20:25.993+0000 7fb55f4a6f40  1 bdev(0x55bafd245180 /var/lib/ceph/osd/ceph-9/block) open path /var/lib/ceph/osd/ceph-9/block
debug 2021-04-04T08:20:25.993+0000 7fb55f4a6f40 -1 bdev(0x55bafd245180 /var/lib/ceph/osd/ceph-9/block) open open got: (13) Permission denied
debug 2021-04-04T08:20:25.993+0000 7fb55f4a6f40  0 osd.9:0.OSDShard using op scheduler ClassedOpQueueScheduler(queue=WeightedPriorityQueue, cutoff=196)
debug 2021-04-04T08:20:25.993+0000 7fb55f4a6f40  0 osd.9:1.OSDShard using op scheduler ClassedOpQueueScheduler(queue=WeightedPriorityQueue, cutoff=196)
debug 2021-04-04T08:20:25.993+0000 7fb55f4a6f40  0 osd.9:2.OSDShard using op scheduler ClassedOpQueueScheduler(queue=WeightedPriorityQueue, cutoff=196)
debug 2021-04-04T08:20:25.993+0000 7fb55f4a6f40  0 osd.9:3.OSDShard using op scheduler ClassedOpQueueScheduler(queue=WeightedPriorityQueue, cutoff=196)
debug 2021-04-04T08:20:25.993+0000 7fb55f4a6f40  0 osd.9:4.OSDShard using op scheduler ClassedOpQueueScheduler(queue=WeightedPriorityQueue, cutoff=196)
debug 2021-04-04T08:20:25.997+0000 7fb55f4a6f40 -1 bluestore(/var/lib/ceph/osd/ceph-9/block) _read_bdev_label failed to open /var/lib/ceph/osd/ceph-9/block: (13) Permission denied
debug 2021-04-04T08:20:25.997+0000 7fb55f4a6f40  1 bluestore(/var/lib/ceph/osd/ceph-9) _mount path /var/lib/ceph/osd/ceph-9
debug 2021-04-04T08:20:25.997+0000 7fb55f4a6f40 -1 bluestore(/var/lib/ceph/osd/ceph-9/block) _read_bdev_label failed to open /var/lib/ceph/osd/ceph-9/block: (13) Permission denied
debug 2021-04-04T08:20:25.997+0000 7fb55f4a6f40  1 bdev create path /var/lib/ceph/osd/ceph-9/block type kernel
debug 2021-04-04T08:20:25.997+0000 7fb55f4a6f40  1 bdev(0x55bafd245180 /var/lib/ceph/osd/ceph-9/block) open path /var/lib/ceph/osd/ceph-9/block
debug 2021-04-04T08:20:25.997+0000 7fb55f4a6f40 -1 bdev(0x55bafd245180 /var/lib/ceph/osd/ceph-9/block) open open got: (13) Permission denied
debug 2021-04-04T08:20:25.997+0000 7fb55f4a6f40 -1 osd.9 0 OSD:init: unable to mount object store
debug 2021-04-04T08:20:25.997+0000 7fb55f4a6f40 -1  ** ERROR: osd init failed: (13) Permission denied
```

It's just so weird that I can write to the folder  as `root` and impersonating `ceph` but for some reason the ceph binary can't. There must be something proactively stopping the permission change (via `--setuser` and `--setgroup`). I've started looking around `seccomp`.

**Comment by t3hmrman** [comment]
Found some similar issues, I think this might be a network thing? It seems like everything in the script actually happens *twice*. Everything is going great until the `-1` (return code?) line where it "falls back to the public interface".

https://github.com/rook/rook/issues/3140

**Comment by t3hmrman** [comment]
Oh and re: https://github.com/rook/rook/issues/7497

I can actually trigger that by `su`ing to the `ceph` user inside an OSD pod and trying to start the OSD myself with the same command. I can write to the disk directory, but the script spits out that error exactly when it starts up (after letting me know it doesn't have to switch user because it's not root).

**Comment by t3hmrman** [comment]
Another note for anyone who might look at this who wants to chase the udev angle -- there are already some rules that *seem to apply* (according to `udevadm test`):

```console
root@all-in-one-01 ~ # cat /usr/lib/udev/rules.d/95-ceph-osd-lvm.rules 
# OSD LVM layout example
# VG prefix: ceph-
# LV prefix: osd-
ACTION=="add", SUBSYSTEM=="block", \
  ENV{DEVTYPE}=="disk", \
  ENV{DM_LV_NAME}=="osd-*", \
  ENV{DM_VG_NAME}=="ceph-*", \
  OWNER:="ceph", GROUP:="ceph", MODE:="660"
ACTION=="change", SUBSYSTEM=="block", \
  ENV{DEVTYPE}=="disk", \
  ENV{DM_LV_NAME}=="osd-*", \
  ENV{DM_VG_NAME}=="ceph-*", \
  OWNER="ceph", GROUP="ceph", MODE="660"
```

It doesn't work even with this but it's worth mentioning that these rules are there (technically they wouldn't target partitions but they're there). 

After turning on `provider: host` and `ROOK_HOSTPATH_REQUIRES_PRIVILEGED: true` and re-installing from scratch, the OSD pods worked just fine (as they did before). After that I did a restart and they were broken again so it looks like the problem is not with the network and/or privileges and is something deeper.

At this point I guess I have to start digging into the Ceph code to find out what interface it's trying to use and why it looks like it prepares everything *then* loops and does the wrong thing..

**Comment by t3hmrman** [comment]
OK so this is definitely the wrong way, but I figured out a way that works -- I removed the `setuser` and `setgroup` arguments from the OSD pods, and they worked immediately. On disk, the files are still owned by UID/GID 167, so I tried `runAsUser: 167` and that did *not* work.

Removing `setuser` and `setgroup` instantly fixed all the osd pods, and I'm getting logs like normal. I'm not sure exactly what's going wrong with `setuid`/`setgid` or the underlying mechanics here but I'm taking the working cluster for now:

If I look at `/etc/passwd`, the `ceph` user *is* mapped to a very different uid:

```console
ceph:x:64045:64045:Ceph storage service:/var/lib/ceph:/usr/sbin/nologin
```

That might actually be the problem here, so I gave it a shot and changed the `ceph` user's uid and gid:

```console
root@all-in-one-01 ~ # groupmod -g 167 ceph
root@all-in-one-01 ~ # usermod -u 167 -g 167 ceph
root@all-in-one-01 ~ # id ceph
uid=167(ceph) gid=167(ceph) groups=167(ceph)
```

After this I reversed the changes (so put `--setuser` and `--setgroup` back in all the OSD deployments), and that did not work. I did notice however that now the files on disk in `/var/lib/rook/rook-ceph` (the disks and metadata) actually have the proper user/group now the `ceph` user. 

I also tried using `runAsUser: 167` and `runAsGroup: 167`, which didn't work either.

Very confused on why the `ceph` user isn't working, but at this point, removing `--setuser` and `--setgroup` will be my workaround.




```

**Comment by galexrt** [comment]
@t3hmrman Do you have any `ceph*` packages installed on your nodes? E.g., `dpkg --list ceph*` or so.

Post the list and / or go ahead and remove the packages from the nodes.

**Comment by t3hmrman** [comment]
Ah I definitely do -- I pre-installed `ceph` and `ceph-mds` since it's Ubuntu 20.04 -- Am I seeing the collision of the `apt` installed `ceph`-created user and Rook's?

I'm in the middle of that build (it's ansible so it's a bit slow) so I will try that as well.

**Comment by galexrt** [comment]
@t3hmrman You don't need to install any of the `ceph*` packages on the nodes as they would conflict with the Ceph user ids of the CentOS Ceph base image conflict with the Ubuntu / Debian packages unless you pre-create the users (which I wouldn't recommend..).

Any reason you install them on the nodes at the moment?

**Comment by t3hmrman** [comment]
Nope -- I looked and I replaced my `ceph-volume lvm zap` commands with `sgdisk` commands a while back, so I'm actually not using the packages at all.

I don't particularly need the packages, so I'll take them out once the build is done -- was this a part of the documentation I overlooked? If not, maybe updating it would be good because I don't feel like I've seen this discussed before, I've looked through what feels like quite a few tickets (I hit an absurd amount of edge cases installing Rook this time around) and I don't think I've seen the collision possibility mentioned before.

**Comment by t3hmrman** [comment]
OK closing this for now -- if someone runs into this they'll probably see this ticket if they search for the exact phrases.

**Comment by sdeoras** [comment]
hi folks, i am running into possible similar issue where `rook-ceph-osd` pod is in crashloop after node reboot. i am happy to start from scratch and wondering if there is a recommendation on what needs to change in the yaml files so node-reboot goes fine. I am using `v1.6.1` of https://github.com/rook/rook.git on a single node ARM64 cluster spawned via k0s. Thanks.

```
$ kubectl --namespace=rook-ceph logs -f pods/rook-ceph-osd-1-f7cdbf88d-vckqg
debug 2021-05-11T23:23:02.804+0000 ffff81f61040  0 set uid:gid to 167:167 (ceph:ceph)
debug 2021-05-11T23:23:02.804+0000 ffff81f61040  0 ceph version 15.2.11 (e3523634d9c2227df9af89a4eac33d16738c49cb) octopus (stable), process ceph-osd, pid 1
debug 2021-05-11T23:23:02.804+0000 ffff81f61040  0 pidfile_write: ignore empty --pid-file
debug 2021-05-11T23:23:02.804+0000 ffff81f61040 -1 bluestore(/var/lib/ceph/osd/ceph-1/block) _read_bdev_label failed to open /var/lib/ceph/osd/ceph-1/block: (13) Permission denied
debug 2021-05-11T23:23:02.804+0000 ffff81f61040 -1  ** ERROR: unable to open OSD superblock on /var/lib/ceph/osd/ceph-1: (2) No such file or directory
```

**Comment by travisn** [comment]
@sdeoras This sounds like #7878, which has a fix in progress

**Comment by sdeoras** [comment]
@travisn i just tested v1.6.3 on arm64 based cluster running k0s and all pods in `rook-ceph` namespace came back fine after reboot! things are looking very good so far, thanks :+1:

**Comment by sdeoras** [comment]
@travisn I am observing the same error with v1.6.3. if you have suggestions on anything i can try to debug, i'll be happy to look further into it and provide more details.

also sorry to piggyback on this thread. if there is another forum i should post this on, pl. advise. thanks.

```
└─ $ ▶ kubectl --namespace=rook-ceph get pods 
NAME                                              READY   STATUS             RESTARTS   AGE
csi-rbdplugin-g54vc                               3/3     Running            12         9d
rook-ceph-operator-65965c66b5-89fh5               1/1     Running            6          9d
csi-cephfsplugin-qcjl7                            3/3     Running            12         9d
rook-ceph-crashcollector-rpi4-0-b86b8b957-99wrj   1/1     Running            4          9d
rook-ceph-mgr-a-857d4665bb-jhdtd                  1/1     Running            15         9d
rook-ceph-osd-prepare-rpi4-0-4qd8w                0/1     Completed          0          7h37m
rook-ceph-osd-0-7f8dfb4945-pt6dq                  1/1     Running            626        9d
csi-rbdplugin-provisioner-857d65496c-b2wzr        6/6     Running            182        9d
csi-cephfsplugin-provisioner-59499cbcdd-skdlp     6/6     Running            173        9d
rook-ceph-mon-a-67445c9f9c-fjqp9                  1/1     Running            319        9d
rook-ceph-osd-1-58c77dfdd8-vs5mj                  0/1     CrashLoopBackOff   218        19h

└─ $ ▶ kubectl --namespace=rook-ceph logs -f pods/rook-ceph-osd-1-58c77dfdd8-vs5mj
debug 2021-06-02T21:24:04.021+0000 ffff99b33040  0 set uid:gid to 167:167 (ceph:ceph)
debug 2021-06-02T21:24:04.021+0000 ffff99b33040  0 ceph version 15.2.11 (e3523634d9c2227df9af89a4eac33d16738c49cb) octopus (stable), process ceph-osd, pid 1
debug 2021-06-02T21:24:04.021+0000 ffff99b33040  0 pidfile_write: ignore empty --pid-file
debug 2021-06-02T21:24:04.021+0000 ffff99b33040 -1 bluestore(/var/lib/ceph/osd/ceph-1/block) _read_bdev_label failed to open /var/lib/ceph/osd/ceph-1/block: (13) Permission denied
debug 2021-06-02T21:24:04.021+0000 ffff99b33040 -1  ** ERROR: unable to open OSD superblock on /var/lib/ceph/osd/ceph-1: (2) No such file or directory
```

**Comment by leseb** [comment]
> @travisn I am observing the same error with v1.6.3. if you have suggestions on anything i can try to debug, i'll be happy to look further into it and provide more details.
> 
> also sorry to piggyback on this thread. if there is another forum i should post this on, pl. advise. thanks.
> 
> ```
> └─ $ ▶ kubectl --namespace=rook-ceph get pods 
> NAME                                              READY   STATUS             RESTARTS   AGE
> csi-rbdplugin-g54vc                               3/3     Running            12         9d
> rook-ceph-operator-65965c66b5-89fh5               1/1     Running            6          9d
> csi-cephfsplugin-qcjl7                            3/3     Running            12         9d
> rook-ceph-crashcollector-rpi4-0-b86b8b957-99wrj   1/1     Running            4          9d
> rook-ceph-mgr-a-857d4665bb-jhdtd                  1/1     Running            15         9d
> rook-ceph-osd-prepare-rpi4-0-4qd8w                0/1     Completed          0          7h37m
> rook-ceph-osd-0-7f8dfb4945-pt6dq                  1/1     Running            626        9d
> csi-rbdplugin-provisioner-857d65496c-b2wzr        6/6     Running            182        9d
> csi-cephfsplugin-provisioner-59499cbcdd-skdlp     6/6     Running            173        9d
> rook-ceph-mon-a-67445c9f9c-fjqp9                  1/1     Running            319        9d
> rook-ceph-osd-1-58c77dfdd8-vs5mj                  0/1     CrashLoopBackOff   218        19h
> 
> └─ $ ▶ kubectl --namespace=rook-ceph logs -f pods/rook-ceph-osd-1-58c77dfdd8-vs5mj
> debug 2021-06-02T21:24:04.021+0000 ffff99b33040  0 set uid:gid to 167:167 (ceph:ceph)
> debug 2021-06-02T21:24:04.021+0000 ffff99b33040  0 ceph version 15.2.11 (e3523634d9c2227df9af89a4eac33d16738c49cb) octopus (stable), process ceph-osd, pid 1
> debug 2021-06-02T21:24:04.021+0000 ffff99b33040  0 pidfile_write: ignore empty --pid-file
> debug 2021-06-02T21:24:04.021+0000 ffff99b33040 -1 bluestore(/var/lib/ceph/osd/ceph-1/block) _read_bdev_label failed to open /var/lib/ceph/osd/ceph-1/block: (13) Permission denied
> debug 2021-06-02T21:24:04.021+0000 ffff99b33040 -1  ** ERROR: unable to open OSD superblock on /var/lib/ceph/osd/ceph-1: (2) No such file or directory
> ```

What are the permissions on the block device?

**Comment by sdeoras** [comment]
@leseb do you mean this?
```
$ ls -la /dev/sda
brw-rw---- 1 167 167 8, 0 Jun  3 00:29 /dev/sda
```
everything was working fine and dynamic volume creation also worked well. as i reported earlier in this thread, the node reboot was also no longer an issue w/ v1.6.3. i am not sure at which point it started throwing the error "/var/lib/ceph/osd/ceph-1: No such file or directory"

**Comment by TheGr8Nik** [comment]
I had the same error, I figured out that I have to change `UID` and `GID` of the `ceph` user in the `/etc/passwd` file to match the values expected by the kubernetes ceph user, in my case are `167:167`.
The change was from

    ceph:x:64045:64045:Ceph storage service:/var/lib/ceph:/usr/sbin/nologin

to

    ceph:x:167:167:Ceph storage service:/var/lib/ceph:/usr/sbin/nologin

**Comment by 24king** [comment]
for me there is no ceph package preinstall and get the same error? on host there is no /var/lib/ceph directory

```
debug 2021-09-30T02:54:32.599+0000 7f908d837f40 -1 ** ERROR: unable to open OSD superblock on /var/lib/ceph/osd/ceph-4: (2) No such file or directory​
```

**Comment by krim404** [comment]
I just want to give the information, that i have the exact same issue on a rollout with cephadm and a fresh debian 11 install with docker as engine. But this time it is a little bit more tricky, as the problem started after a reboot long after the creation of the OSDs (and placing data on it). 

after changing the uid of the debian maintained ceph user from 64045:64045 to 167:167 it will start up again.
problem is the permission inside of the container of the /dev/dm-xy, which is only user accessable to 64045.

Issue only affects the osd container, all other services will work. Strange thing is: debian 10 has the exact same user ids, but there it works without any problem.

**Comment by javdrher** [comment]
@krim404 did you find a solution for this? We encounter the same. Maybe a reboot fixes the /dev/dm-xy permission?

**Comment by krim404** [comment]
yes, we fixed it by updating the permissions, but i dont remember anymore how, only that there was a very easy fix.

**Comment by szczepl** [comment]
I have this in ceph 17.2.6 (rook version 1.11.4). For me, solution was to create debug pod for container 'activate' and do the chmod 777 /dev/sdb (which is linked from /var/lib/ceph/ceph-1/block in my case). Immediately after that container osd started to work. Of course that's 'dirty hack' but this is the only thing which worked for me.
